cache/cache.go

```diff
@@ -0,0 +1,105 @@
+package cache
+
+import (
+	"context"
+	"log"
+	"os"
+	"path"
+
+	"github.com/samsarahq/taskrunner"
+	"github.com/samsarahq/taskrunner/shell"
+)
+
+const CachePath = ".cache/taskrunner"
+
+var CacheDir = path.Join(os.Getenv("HOME"), CachePath)
+
+type Cache struct {
+	ranOnce     map[*taskrunner.Task]bool
+	snapshotter *snapshotter
+	cacheFile   string
+	allDirty    bool
+	dirtyFiles  []string
+}
+
+func New() *Cache {
+	return &Cache{
+		ranOnce:   make(map[*taskrunner.Task]bool),
+		cacheFile: path.Join(CacheDir, "example.json"),
+	}
+}
+
+func (c *Cache) Start(opt shell.RunOption) error {
+	c.snapshotter = newSnapshotter(
+		context.TODO(),
+		func(ctx context.Context, command string, opts ...shell.RunOption) error {
+			return shell.Run(ctx, command, append(opts, opt)...)
+		},
+	)
+
+	s, err := c.snapshotter.read(c.cacheFile)
```

Comment:
since these are called outside of the struct, can they be exported methods? (diff and write as well)

Reply:
For sure. I agree with the approach. I wish privacy actually worked that way in Go.